collections/src/main/scala/strawman/collection/mutable/AnyRefMap.scala

```diff
@@ -0,0 +1,490 @@
+      else if (imbalanced) repack()
+      None
+    }
+    else {
+      val ans = Some(_values(i).asInstanceOf[V])
+      _hashes(i) = h
+      _keys(i) = k
+      _values(i) = value.asInstanceOf[AnyRef]
+      ans
+    }
+  }
+
+  /** Updates the map to include a new key-value pair.
+   *
+   *  This is the fastest way to add an entry to an `AnyRefMap`.
+   */
+  override def update(key: K, value: V): Unit = {
+    val h = hashOf(key)
+    val k = key
+    val i = seekEntryOrOpen(h, k)
+    if (i < 0) {
+      val j = i & IndexMask
+      _hashes(j) = h
+      _keys(j) = k
+      _values(j) = value.asInstanceOf[AnyRef]
+      _size += 1
+      if ((i & VacantBit) != 0) _vacant -= 1
+      else if (imbalanced) repack()
+    }
+    else {
+      _hashes(i) = h
+      _keys(i) = k
+      _values(i) = value.asInstanceOf[AnyRef]
+    }
+  }
+
+  /** Adds a new key/value pair to this map and returns the map. */
+  def add(key: K, value: V): this.type = { update(key, value); this }
+
+  def add(kv: (K, V)): this.type = { update(kv._1, kv._2); this }
+
+  def subtract(key: K): this.type = {
+    val i = seekEntry(hashOf(key), key)
+    if (i >= 0) {
+      _size -= 1
+      _vacant += 1
+      _hashes(i) = Int.MinValue
+      _keys(i) = null
+      _values(i) = null
+    }
+    this
+  }
+
+  def iterator(): Iterator[(K, V)] = new Iterator[(K, V)] {
+    private[this] val hz = _hashes
+    private[this] val kz = _keys
+    private[this] val vz = _values
+
+    private[this] var index = 0
+
+    def hasNext: Boolean = index<hz.length && {
+      var h = hz(index)
+      while (h+h == 0) {
+        index += 1
+        if (index >= hz.length) return false
+        h = hz(index)
+      }
+      true
+    }
+
+    def next(): (K, V) = {
+      if (hasNext) {
+        val ans = (kz(index).asInstanceOf[K], vz(index).asInstanceOf[V])
+        index += 1
+        ans
+      }
+      else throw new NoSuchElementException("next")
+    }
+  }
+
+  override def foreach[U](f: ((K,V)) => U): Unit = {
+    var i = 0
+    var e = _size
+    while (e > 0) {
+      while(i < _hashes.length && { val h = _hashes(i); h+h == 0 && i < _hashes.length}) i += 1
+      if (i < _hashes.length) {
+        f((_keys(i).asInstanceOf[K], _values(i).asInstanceOf[V]))
+        i += 1
+        e -= 1
+      }
+      else return
+    }
+  }
+
+  override def clone(): AnyRefMap[K, V] = {
+    val hz = java.util.Arrays.copyOf(_hashes, _hashes.length)
+    val kz = java.util.Arrays.copyOf(_keys, _keys.length)
+    val vz = java.util.Arrays.copyOf(_values,  _values.length)
+    val arm = new AnyRefMap[K, V](defaultEntry, 1, false)
+    arm.initializeTo(mask, _size, _vacant, hz, kz,  vz)
+    arm
+  }
+
+  override def concat[V2 >: V](xs: strawman.collection.Iterable[(K, V2)]): AnyRefMap[K, V2] = {
+    val arm = clone().asInstanceOf[AnyRefMap[K, V2]]
+    xs.foreach(kv => arm += kv)
+    arm
+  }
+
+  /*override*/ def updated[V1 >: V](key: K, value: V1): AnyRefMap[K, V1] = {
```

Comment:
Indeed we don’t have an `updated` operation anymore on mutable Maps. We should probably deprecate that one, what do you think?

Reply:
Is there a good replacement? `updated` can widen the value type which otherwise requires a cast. The scalac codebase doesn't use the method though.

Reply:
The replacement is `myMutableMap.toMap.updated(k, v)`

Reply:
This would never be a suitable replacement because it's horribly slow. If you didn't care about performance you wouldn't be using an `AnyRefMap` in the first place.

Reply:
I think I've found a solution: I'll add `AnyRefMap.from`. The object can't implement any of the standard factories, so this method doesn't exist yet, but it's a natural place to perform a widening copy and we can easily optimize the case where you're copying an existing `AnyRefMap`. With this method `m.updated[V1](k,v)` becomes `{ val m2 = AnyRefMap.from[K, V1](m); m2 += ((k,v)); m2 }`.